packages/core/upload/server/services/extensions/content-manager/entity-manager.js

```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+const { mapAsync, traverseEntity } = require('@strapi/utils');
+const { getService } = require('../../../utils');
+
+/**
+ * Visitor function to sign media URLs
+ * @param {Object} schema
+ * @param {string} schema.key - The key of the attribute
+ * @param {string} schema.value - The value of the attribute
+ * @param {Object} schema.attribute - The attribute definition
+ * @param {Object} entry
+ * @param {Function} entry.set - The set function to update the value
+ */
+const signEntityMediaVisitor = async ({ key, value, attribute }, { set }) => {
+  const { signFileUrls } = getService('file');
+
+  if (!value || attribute.type !== 'media') {
+    return;
+  }
+
+  // If the attribute is repeatable sign each file
+  if (attribute.multiple) {
+    const signedFiles = await mapAsync(value, signFileUrls);
+    set(key, signedFiles);
```

Comment:
Should we return values instead of using a function that set the value ? 🤔 Isn't this hard to maintain / read / understand to not return anything and call sort of a callback?

Reply:
I personally find this much easier to follow, as you can set but also remove fields (or more actions)

This is how the traverse entity util is defined in strapi and we use it for sanitizing and a couple more places. We are starting discussions to refactor it, so we could see how we can improve it together!